ashlar/thumbnail.py

```diff
@@ -0,0 +1,101 @@
+from __future__ import print_function, division
+import sys
+try:
+    import pathlib
+except ImportError:
+    import pathlib2 as pathlib
+import numpy as np
+from . import utils
+from skimage.transform import rescale
+from skimage.feature import register_translation
+from skimage.io import imsave
+import warnings
+
+
+def make_thumbnail(reader, channel=0, scale=0.05):
+    
+    metadata = reader.metadata
+    positions = metadata.positions - metadata.origin
+    coordinate_max = (positions + metadata.size).max(axis=0) 
+    mshape = ((coordinate_max+ 1) * scale).astype(int)
+    mosaic = np.zeros(mshape, dtype=np.uint16)
+
+    total = reader.metadata.num_images
+    for i in range(total):
+        sys.stdout.write("\r        Loading %d/%d" % (i + 1, total))
+        sys.stdout.flush()
+        utils.paste(
+            mosaic, rescale(reader.read(c=channel, series=i), scale), positions[i] * scale, np.maximum
+        )
+    print()
+
+    return mosaic
+
+
+def calculate_image_offset(img1, img2, upsample_factor=1):
+
+    ref = utils.fft2(utils.whiten(img1, 0))
+    test = utils.fft2(utils.whiten(img2, 0))
+
+    shift, error, _ = register_translation(ref, test, upsample_factor, 'fourier')
+
+    return shift
+
+
+def calculate_cycle_offset(reader1, reader2, channel=0, scale=0.05, save=(False, False)):
+    
+    img1 = reader1.thumbnail_img \
```

Comment:
Computing the thumbnail here seems a little awkward. It might make more sense to compute it in the `run` method of the aligners (like `EdgeAligner` already does) and store it in the `.thumbnail` attribute of the aligner objects. Then this function's arguments would become `(aligner1, aligner2, scale)` (assuming we drop the image saving code as suggested below) and this if/else logic goes away.